macro/vignettes/design_meta_timestep.Rmd

```diff
@@ -0,0 +1,403 @@
+From `r system("git config --get remote.origin.url", intern = TRUE)`
+on `r date()`, generated by `r Sys.info()[["effective_user"]]`.
+
+```{r load_libraries}
+library(macro)
+```
+
+
+# Overview
+
+I'd like to be able to build a time step quickly from a specification, at least for some
+simple models. In this case, I'm thinking about specifying linear equations with coefficients
+that are constant across a time step.
+$$
+\frac{dx}{dt}=Ax=\left[\begin{array}{cc}-h & \sin(r) \\ h & -\sin(r)\end{array}\right]x
+$$
+I'm using a $\sin$ there to make the point that the rates can be arbitrary functions of
+parameters. We could write that equation as
+```{r tease_apart}
+si_equations <- list(
+U ~ -h * U + sin(r) * X,
+X ~ h * U - sin(r) * X
+)
+```
+I'll assume this is a differential form of an ODE that reflects a master equation.
+That is, I'll assume that everything that leaves one compartment enters another compartment
+and that there are no constant terms.
+
+**Are these limitations the best choices?**
+
+From this kind of specification, we can produce lots of different kinds of time steps.
+
+# Discrete Time Steps
+
+## The Matrix for the Master Equation
+Most of them are simple permutations of the specification. The simplest is a function
+that returns what it received, the matrix $A$. We can look at the text of the function itself, thanks
+to R's ability to print the body of a function.
+```{r example_cme}
+(si_cme <- cme_maker(si_equations))
+```
+This function executes, given a parameter list.
+```{r example_cme_execute}
+si_cme(list(h = 0.4, r = 0.1))
+```
+How would you use this? You could pass this function to an ODE solver
+in order to get change across a step with constant parameters.
+
+## Small Time Step
+For small time steps, we approximate the time step as $I-A t$.
+$$
+\left[\begin{array}{cc}1-ht & \sin(r)t \\ ht & 1-\sin(r)t\end{array}\right]
+$$
+
+
+```{r create_small}
+(si_small <- small_step_maker(si_equations))
+```
+We can execute it by passing in the original parameters, plus a time.
+```{r run_small}
+si_small(list(h = 0.4, r = 0.1, t = 0.05))
+```
+
+How do you use this?
+The small time step is a very common way to build steps. We think
+of rate times time as the marginal probability for a jump in a Markov
+chain. If parameters keep changing, then this function can build the matrix quickly.
+If you wanted to simulate a long interval that has constant
+parameters, you could split it into $2^(2n)$ steps and square this matrix
+$n$ times.
+
+When rates become large, the small time step can become not only erroneous
+but also infeasible, because the rate to remain in a compartment is
+one minus the sum of rates to leave, so it can become negative.
+
+## Exact Exponential
+
+It is possible to construct the rate matrix, $At$, and exponentiate it, in order
+to get an exact solution to the time step. The time step is a numerical calculation
+of the expontial in
+$$
+x_{t+1} = e^{At} x_t.
+$$
+It's as good as integrating, and R
```

Comment:
I might be even a little more pedantic here and say "It's as good as integrating the master equation (ODEs) component by component" or something like that.